### PR TITLE
chunked: skip file metadata for composefs-like links

### PR DESCRIPTION
**Pull request by giuseppe**
if the file is created using the object-store flat directory format, there is no need to set its inodes attributes, as anyway they are ignored when creating the composefs binary blob.

Similarly, we do not need to set file attributes when the deduplication happened through a hard link.

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
@mtrmac @alexlarsson @cgwalters PTAL

File: pkg/chunked/internal/compression.go [hunk 1]
```diff
@@ -51,6 +51,10 @@ type FileMetadata struct {
 
 	// internal: computed by mergeTOCEntries.
 	Chunks []*FileMetadata `json:"-"`
+	// SkipSetAttrs is set when the file attributes must not be
+	// modified, e.g. it is a hard link from a different source,
+	// or a composefs file.
+	SkipSetAttrs bool
```
Comment (mtrmac) [inline pkg/chunked/internal/compression.go:57]: Is this really intended to be maintained in the TOC?

If not, can this field be tracked internally inside `pkg/chunked` only?

**Comment by cgwalters** [comment]
Apart from the other comment, this seems sane; are there no integration tests that run with composefs enabled on this repo?  It seems like it wouldn't be too hard to add a test for this if so.  (And if not, probably need integration coverage for chunked/composefs here)

**Review by mtrmac — COMMENTED** [review]
BTW, the code to check whether the hardlink destination matches the expected attributes (`canDedupFileWithHardLink`?) and this code to set/not set the attributes are way too distant for me to be comfortable; I’m not smart enough to believe I would be able to keep the two in sync long-term.

File: pkg/chunked/internal/compression.go (hunk 1 above)
Comment (giuseppe) [reply pkg/chunked/internal/compression.go]: yeah, let's clean this up and remove it from the TOC definition.

What do you think about the new version?

**Comment by giuseppe** [comment]
> Apart from the other comment, this seems sane; are there no integration tests that run with composefs enabled on this repo? It seems like it wouldn't be too hard to add a test for this if so. (And if not, probably need integration coverage for chunked/composefs here)

this is still something we miss, as we don't have any test for composefs

File: pkg/chunked/storage_linux.go [hunk 2]
```diff
@@ -58,6 +58,20 @@ const (
 	copyGoRoutines = 32
 )
 
+// fileMetadata is a wrapper around internal.FileMetadata with additional private fields that
+// are not part of the TOC document.
+type fileMetadata struct {
+	internal.FileMetadata
+
+	// Chunks are computed by mergeTOCEntries.
+	Chunks []*internal.FileMetadata `json:"-"`
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:67]: - The internal fields can be package-private
- I read the the `json` annotation to suggest that this struct can be marshaled/unmarshaled; that’s AFAICS not the case, so dropping the annotation would make this easier to follow.

File: pkg/chunked/storage_linux.go [hunk 3]
```diff
@@ -58,6 +58,20 @@ const (
 	copyGoRoutines = 32
 )
 
+// fileMetadata is a wrapper around internal.FileMetadata with additional private fields that
+// are not part of the TOC document.
+type fileMetadata struct {
+	internal.FileMetadata
+
+	// Chunks are computed by mergeTOCEntries.
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:66]:
```suggestion
	// If FileMetadata.Type == TypeReg, this stores the TypeChunk entries relevant to this entry.
```
(“relevant to” can probably be improved upon)

File: pkg/chunked/storage_linux.go [hunk 4]
```diff
@@ -582,7 +601,10 @@ func (o *originFile) OpenFile() (io.ReadCloser, error) {
 }
 
 // setFileAttrs sets the file attributes for file given metadata
-func setFileAttrs(dirfd int, file *os.File, mode os.FileMode, metadata *internal.FileMetadata, options *archive.TarOptions, usePath bool) error {
+func setFileAttrs(dirfd int, file *os.File, mode os.FileMode, metadata *fileMetadata, options *archive.TarOptions, usePath bool) error {
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:604]: (Unrelated: if I’m not mistaken, `mode` is always `os.FileMode(metadata.Mode)`, so the parameter could be eliminated throughout the call stack. Perhaps best to do in a separate PR.)

File: pkg/chunked/storage_linux.go [hunk 5]
```diff
@@ -1729,9 +1752,16 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 
 	var missingParts []missingPart
 
-	output.UIDs, output.GIDs = collectIDs(toc.Entries)
+	var entries []fileMetadata
+	for _, e := range toc.Entries {
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:1756]: (This works just fine; alternatively, maybe it could be the job of `mergeTocEntries` to convert from the one type to the other.

That way, the semantics could be separated by type: `[]internal.FileMetadata` can contain top-level `TypeChunk` entries, and `[]fileMetadata` never contains top-level `TypeChunk`, they are all collected in `.Chunks`.)

File: pkg/chunked/storage_linux.go [hunk 6]
```diff
@@ -58,6 +58,20 @@ const (
 	copyGoRoutines = 32
 )
 
+// fileMetadata is a wrapper around internal.FileMetadata with additional private fields that
+// are not part of the TOC document.
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:62]:
```suggestion
// are not part of the TOC document.
// Type: TypeChunk entries are stored in Chunks, the primary [fileMetadata] entries never use TypeChunk
```

File: pkg/chunked/storage_linux.go [hunk 7]
```diff
@@ -1987,6 +2017,10 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 
 			switch chunk.ChunkType {
 			case internal.ChunkTypeData:
+				fm := fileMetadata{
+					FileMetadata: *chunk,
+				}
+				chunk := &fm
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:2023]: Following the principle that `fileMetadata` entries only contain `TypeChunk` entries in the `.Chunks` field, not at the top level, it seems to me that the type of the `findChunkInOtherLayers` parameter could remain `internal.FileMetadata`, so that we don’t need to create this usually-invalid entry.

**Review by mtrmac — COMMENTED** [review]
This is nice.

Apart from comment around `internal.ChunkTypeData` (to consistently never use the internal `fileMetadata` for `TypeChunk` values), all others are just tentative aesthetic suggestions.

**Comment by giuseppe** [comment]
thanks, took care of the comments (except https://github.com/containers/storage/pull/1862#discussion_r1530966629 ) and pushed a new version

File: pkg/chunked/storage_linux.go [hunk 8]
```diff
@@ -2045,9 +2068,16 @@ func mustSkipFile(fileType compressedFileType, e internal.FileMetadata) bool {
 	return false
 }
 
-func (c *chunkedDiffer) mergeTocEntries(fileType compressedFileType, entries []internal.FileMetadata) ([]internal.FileMetadata, int64, error) {
+func (c *chunkedDiffer) mergeTocEntries(fileType compressedFileType, tocEntries []internal.FileMetadata) ([]fileMetadata, int64, error) {
 	var totalFilesSize int64
 
+	var entries []fileMetadata
+	for _, e := range tocEntries {
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:2075]: AFAICS this copy/allocation would also be cheap to eliminate: We are already building `mergedEntries` as an array of the right type, so…

File: pkg/chunked/storage_linux.go [hunk 9]
```diff
@@ -2069,7 +2099,7 @@ func (c *chunkedDiffer) mergeTocEntries(fileType compressedFileType, entries []i
 		}
 	}
 
-	mergedEntries := make([]internal.FileMetadata, size)
+	mergedEntries := make([]fileMetadata, size)
 	m := 0
 	for i := 0; i < len(entries); i++ {
 		e := entries[i]
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:2105]: …
```go
    e := fileMetadata{ FileMetadata: entries[i] }
```
might be all that is necessary.

**Review by mtrmac — COMMENTED** [review]
Thanks!

LGTM, feel free to merge as is or add that small cleanup.
